src/kubernetes.cr

```diff
@@ -23,27 +23,50 @@ module Kubernetes
     end
 
     def self.from_config(config : Config, *, context context_name : String = config.current_context)
-      if context_entry = config.contexts.find { |c| c.name == context_name }
-        if cluster_entry = config.clusters.find { |c| c.name == context_entry.context.cluster }
-          if user_entry = config.users.find { |u| u.name == context_entry.context.user }
-            file = File.tempfile prefix: "kubernetes", suffix: ".crt" do |tempfile|
-              Base64.decode cluster_entry.cluster.certificate_authority_data, tempfile
-            end
-            at_exit { file.delete }
+      context_entry = config.contexts.find { |c| c.name == context_name }
+      if !context_entry
+        raise ArgumentError.new("No context #{context_name.inspect} found in Kubernetes config")
+      end
 
-            new(
-              server: cluster_entry.cluster.server,
-              certificate_file: file.path,
-              token: user_entry.user.credential.status.token,
-            )
-          else
-            raise ArgumentError.new("No user #{context_entry.context.user.inspect} found in Kubernetes config")
-          end
-        else
-          raise ArgumentError.new("No cluster #{context_entry.context.cluster.inspect} found in Kubernetes config")
+      cluster_entry = config.clusters.find { |c| c.name == context_entry.context.cluster }
+      if !cluster_entry
+        raise ArgumentError.new("No cluster #{context_entry.context.cluster.inspect} found in Kubernetes config")
+      end
+
+      user_entry = config.users.find { |u| u.name == context_entry.context.user }
+      if !user_entry
+        raise ArgumentError.new("No user #{context_entry.context.user.inspect} found in Kubernetes config")
+      end
+
+      file = File.tempfile prefix: "kubernetes", suffix: ".crt" do |tempfile|
+        Base64.decode cluster_entry.cluster.certificate_authority_data, tempfile
+      end
+      at_exit { file.delete }
+
+      user = user_entry.user
+
+      if user.client_certificate_data && user.client_key_data
+        client_cert_file = File.tempfile prefix: "kubernetes", suffix: ".crt" do |tempfile|
+          Base64.decode user.client_certificate_data.not_nil!, tempfile
+        end
+        private_key_file = File.tempfile prefix: "kubernetes", suffix: ".crt" do |tempfile|
+          Base64.decode user.client_key_data.not_nil!, tempfile
         end
+        at_exit { client_cert_file.delete; private_key_file.delete }
+
+        new(
+          token: "",
+          server: cluster_entry.cluster.server,
+          certificate_file: file.path,
+          client_cert_file: client_cert_file.path,
+          private_key_file: private_key_file.path,
+        )
       else
-        raise ArgumentError.new("No context #{context_name.inspect} found in Kubernetes config")
+        new(
+          server: cluster_entry.cluster.server,
+          certificate_file: file.path,
+          token: user_entry.user.credential.status.token,
+        )
```

Comment:
Good catch. As I remember, it should always go together: public and private keys. So, the public key is sent to the server to initialize the TLS handshake. Then, the response from the server is verified (or decrypted) if it was signed with the public key, by using the private key.

My understanding with public key a message is encrypted and with private it could be decrypted.